### ISSUE TITLE
Add debugging tools to admin bar

### ISSUE THREAD

**Issue by schlessera** [ampproject/amp-wp#3588]
## Feature description

As debugging is a big part of working with the AMP plugin, I think we need to ensure we provide as many tools as possible to make this as easy as it can be.

Right now, debugging anything that happens during the shutdown handler can only be properly inspected through XDebug or logging. The most straight-forward way that most developers use to get quick info (dumping) is not accessible at that point.

I'd like to add a debugging toolset to the AMP admin bar integration (maybe hidden behind a `AMP_DEBUG` switch?) that shows meaningful diagnostic data, and most of all, provides a container to dump data into.

A developer should be able to do something like `AMP_Debug::dump( $some_variable );` and it would populate a submenu in the admin bar with entries.

Each dump would show how many times it was triggered, and what the individual values were.

Cheap example menu structure diagram thingy:

```
AMP (top-level menu entry)
   Re-validate
   View non-AMP version
   ---
   CSS Size
      [...]
   Dumps
      $some-variable [6]
         (bool) true
         (bool) true
         (bool) true
         (bool) true
         (bool) true
         (bool) false
```

This would come in extremely handy for getting a quick overview of what a certain piece of code does, and could even be used on production sites (😬).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by schlessera** [comment]
The idea for this came from the Tracy Debugger, which provides a `bdump()` for dumping into the debugger bar:
https://tracy.nette.org/en/guide#toc-variable-dumping

**Comment by swissspidy** [comment]
I've always been dreaming about a Query Monitor integration... If the solution here is implementation agnostic, that would be huge.

**Comment by schlessera** [comment]
The main work is collecting the data. Whether you then output it as a list in the admin bar or a list in QM is just a matter of picking the right classes...?

**Comment by swissspidy** [comment]
That's what I wanted to hear, thanks :-)

**Comment by westonruter** [comment]
> CSS Size

See also #2169 and #1801. It would be cool to show a `<meter>` element indicating how close the user is to the 50KB limit before going to the Validated URL screen, and (currently) looking at the source code of the page for the CSS manifest HTML comment.

> A developer should be able to do something like `AMP_Debug::dump( $some_variable );` and it would populate a submenu in the admin bar with entries.

I often dump out variables with this:

```php
error_log( json_encode( $x, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES ) );
```

I then tail the error log. Would this be any different other than avoiding the need to have access to the error log?

**Comment by amedina** [comment]
@schlessera Have you continued exploring this?

**Comment by jwold** [comment]
Is there any UX work that I can support on this, or is it already pretty much thought out?

**Comment by westonruter** [comment]
I think we need to first nail down what information specifically we want to surface from a technical perspective, and then we can explore UX.

**Comment by amedina** [comment]
I agree. Let's think about the feature in the context of the "workflows" we
are discussing as part of the upcoming roadmap. No implementation should be
started on this before we nail down the feature requirements.
